File: 0070-climbing-stairs/0070-climbing-stairs.py
# class Solution:
#     def climbStairs(self, n: int) -> int:
        
class Solution:
    def climbStairs(self, n: int) -> int:
        self.dp = [-1] * (n + 1)
        return self.find(n)

    def find(self, n):
        if n == 0:
            return 1
        if n < 0:
            return 0

        if self.dp[n] != -1:
            return self.dp[n]

        self.dp[n] = self.find(n - 1) + self.find(n - 2)
        return self.dp[n]

File: 0070-climbing-stairs/0070-climbing-stairs.cpp
class Solution {
    vector<int> dp;
public:
    int climbStairs(int n) {
        dp.resize(n+1,-1);
        return find(n);
    }

    int find(int n){
        if (n==0) return 1;
        if (n<1) return 0;
        if (dp[n]!= -1) return dp[n];
        int take1 = find(n-1);
        int take2 = find(n-2);
        return dp[n] = take1 + take2;
    }
};